File: src/day1.hs
import Control.Monad
import AOC

main = aoc "day1" part1 part2

part1 input = increases $ map read input

part2 input = increasesWindow $ map read input

increases :: [Int] -> Int
increases [] = 0
increases (x:[]) = 0
increases (x:y:xs) = (if y > x then 1 else 0) + increases (y:xs)

increasesWindow :: [Int] -> Int
increasesWindow [] = 0
increasesWindow (x:[]) = 0
increasesWindow (x:y:[]) = 0
increasesWindow (x:y:z:[]) = 0
increasesWindow (x:y:z:a:xs) = (if a > x then 1 else 0) + increasesWindow (y:z:a:xs)
--  (y + z + a) > (x + y + z) is eqiuv to a > x

File: src/day1.ml
open AOC

let rec increases = function
| [] -> 0
| (x :: []) -> 0
| (x :: y :: xs) -> (if y > x then 1 else 0) + increases (y :: xs)

let rec increasesWindow = function
| [] -> 0
| (x :: []) -> 0
| (x :: y :: []) -> 0
| (x :: y :: z :: []) -> 0
| (x :: y :: z :: a :: xs) -> (if a > x then 1 else 0) + increasesWindow (y :: z :: a :: xs)

let part1 input = string_of_int (increases (List.map int_of_string input))
let part2 input = string_of_int (increasesWindow (List.map int_of_string input))

let () = aoc "day1" part1 part2
